### PR TITLE
[Object Spilling] Unpin before updating URL

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Currently, we unpin objects "after" the spilled object URL is updated to the owner. This is unnecessary because we don't have any logic that needs to happen after the owner updates the spilled URL to unpin the object (it just GC the objects). 

We also found out that when we check if object is restorable from the local raylet, it checks the OBOD state, which can be stale (and it can be more stale after this PR). We fix it by checking the consistent state from local_object_manager directly that provides the correct information. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/local_object_manager.cc [hunk 1]
```diff
@@ -261,51 +261,26 @@ void LocalObjectManager::SpillObjectsInternal(
                 auto it = objects_pending_spill_.find(object_id);
                 RAY_CHECK(it != objects_pending_spill_.end());
                 pinned_objects_size_ += it->second.first->GetSize();
+                num_bytes_pending_spill_ -= it->second.first->GetSize();
```
Comment (rkooo567) [inline src/ray/raylet/local_object_manager.cc:264]: This was a bug. We have one more bug in stats update path. We can talk about it more @scv119  later

Comment (scv119) [reply src/ray/raylet/local_object_manager.cc:264]: yeah, trying to understand the behavior change of this PR. we can chat on Monday.

**Comment by scv119** [comment]
As we discussed, this might have undesired behavior on the object restoration path.

**Comment by rkooo567** [comment]
Added it! TODO; Fix the broken pull manager test

**Review by scv119 — APPROVED** [review]
mostly nits and questions!

**Pull request merged**